### PR TITLE
reverse the mongo/mysql query logic in some cases [PERF-239; risk: low]

### PR DESCRIPTION
**Pull request by davidangb**
Query logic before this PR:
1. Retrieve all documents from Mongo that match the search criteria. This could be 1 document in the case of get-workflow-detail,  or ~37,000 documents in the case of list-workflows.
2. Retrieve all permissions from MySQL for the current user. This varies by user but 3,000 is a decent ballpark estimation.
3. Intersect the two results to generate a final list.

In the case of get-workflow-detail, where we get 0 or 1 documents from Mongo, the MySQL query is very inefficient: we get ~3,000 permissions to validate one document.

In the case of list-workflows, the Mongo query is wildly inefficient. We retrieve ~37,000 documents and then discard ~90% of them.

As of this PR, query logic is:
1. Inspect the criteria in the request to predict how many Mongo results we'll get.
2. If we expect Mongo results to be small (e.g. get-workflow-detail), query Mongo, then query MySQL to validate permissions for the Mongo results. When Mongo returned < 100 results, use the Mongo results as an IN clause for the MySQL query.
3. If we expect Mongo results to be large (e.g list-workflows), query MySQL first, then use the MySQL results as Mongo's equivalent of an IN clause for the Mongo query.

AgoraPermissions

-----

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Make sure Swagger is updated if API changes
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [ ] Tell the tech lead (TL) that the PR exists if they wants to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Product Owner** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/agora/server/business/AgoraBusiness.scala [hunk 1]
```diff
@@ -410,6 +411,108 @@ class AgoraBusiness(permissionsDataSource: PermissionsDataSource)(implicit ec: E
...
+      // NB: at some point we may need to batch queries to Mongo, if the "$where:" clause containing aliases
+      // becomes too large. Mongo query-document size limit is 16MB and that feels far away, but ...
+      val entityResults = AgoraDao.createAgoraDao(entityTypes)
+        .findByAliases(aliases.toList.distinct, agoraSearch, agoraProjection)
+
+      val rawCount = aliases.length
+      val filteredCount = entityResults.length
+      val efficiency:Float = filteredCount.toFloat/rawCount.toFloat
+
+      logger.info(f"""{"caller": "findByMySqlFirst(${entityTypes.sorted.mkString(",")})", "method": "findByMySqlFirst", "efficiency": $efficiency, "filtered": $filteredCount, "raw": $rawCount}""")
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/agora/server/business/AgoraBusiness.scala]: why `f"` and not `s"` here?

**Review by helgridly — APPROVED** [review]

File: src/main/scala/org/broadinstitute/dsde/agora/server/business/AgoraBusiness.scala (hunk 1 above)
Comment (davidangb) [reply src/main/scala/org/broadinstitute/dsde/agora/server/business/AgoraBusiness.scala]: oh hah that's a copy/paste from elsewhere in the codebase. I had originally intended to `printf`-style format the efficiency into a percentage or maybe cap its significant digits  ...  eventually bailed and just output it as-is

**Comment by davidangb** [comment]
@helgridly I re-requested a review b/c I added a commit with nontrivial changes. Thanks!

**Review by dvoet — APPROVED** [review]

File: src/main/scala/org/broadinstitute/dsde/agora/server/business/AgoraBusiness.scala [hunk 2]
```diff
@@ -410,6 +411,82 @@ class AgoraBusiness(permissionsDataSource: PermissionsDataSource)(implicit ec: E
...
+    }
+
+    if (isCriteriaSpecific(agoraSearch)) {
+      findByMongoFirst(agoraSearch, agoraProjection, entityTypes, username)
+    } else {
+      findByMySqlFirst(agoraSearch, agoraProjection, entityTypes, username)
+    }
+  }
+
+  // TODO: rename method
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/agora/server/business/AgoraBusiness.scala]: not sure if you missed this so i'm noting it

**Review by helgridly — APPROVED** [review]

**Review by helgridly — APPROVED** [review]

File: src/main/resources/reference.conf [hunk 3]
```diff
@@ -37,6 +39,9 @@ akka {
   loggers = ["akka.event.slf4j.Slf4jLogger"]
 }
 
+akka.http.server.idle-timeout = 95 s # Akka default is 60 s
+akka.http.server.request-timeout = 65 s # Akka default is 20 s
```
Comment (dvoet) [inline src/main/resources/reference.conf:43]: I recall something about these settings not working unless they are in agora.conf

**Review by dvoet — APPROVED** [review]

File: src/main/resources/reference.conf (hunk 3 above)
Comment (davidangb) [reply src/main/resources/reference.conf:43]: I'll test this. I verified by outputting the value from inside Agora code ... but that doesn't mean they're actually working. I'll leave them here and add to firecloud-develop if necessary.

**Pull request merged**